qf_lib/plotting/charts/waterfall_chart.py

```diff
@@ -15,61 +15,70 @@
+    def _configure_axis(self):
+        data_element_decorators = self.get_data_element_decorators()
+        indices = [i for index in [d.data.index for d in data_element_decorators] for i in index]
+        self.axes.set_xlim(0, len(indices))
         self.axes.tick_params(axis='both', which='major', labelsize=10)
-        self.axes.set_xticks(range(len(self.data.index) + 2))
-        self.axes.set_xticklabels(['', *self.data.index, ''])
-        self.axes.set_title(self.title, y=1.04)
+        self.axes.set_xticks(range(len(indices) + 2))
+        self.axes.set_xticklabels(['', *indices, ''])
 
-        self._apply_decorators()
+    def _add_text(self):
+        data_element_decorators = self.get_data_element_decorators()
+        self.cumulative_sum = np.cumsum(np.concatenate([d.data.values for d in data_element_decorators]))
+        index = 0
+        for data_element in data_element_decorators:
+            for value in data_element.data.items():
+                y_loc = value[1] if index == 0 or value[0] == self.total_value else self.cumulative_sum[index]
+                text = "{:.2f}%".format(value[1]) if self.percentage else value[1]
+                self.add_decorator(TextDecorator(text, y=DataCoordinate(y_loc + 0.02),
+                                                 x=DataCoordinate(index + 1),
+                                                 verticalalignment='bottom',
+                                                 horizontalalignment='center',
+                                                 fontsize=10))
+                index += 1
 
     def _plot_waterfall(self, index, value):
         if index == 0 or value[0] == self.total_value:
             color = '#A6A6A6' if value[0] == self.total_value else '#4472C4' if value[1] > 0 else '#ED7D31'
             bottom = 0
-            y_loc = value[1]
-
         else:
             color = '#4472C4' if value[1] > 0 else '#ED7D31'
             bottom = self.cumulative_sum[index - 1]
-            y_loc = self.cumulative_sum[index]
-
-        text = "{:.2f}%".format(value[1]) if self.percentage else value[1]
 
         self.axes.bar(index + 1, value[1], bottom=bottom, color=color)
-        self.add_decorator(TextDecorator(text, y=DataCoordinate(y_loc + 0.02), x=DataCoordinate(index + 1),
-                                         verticalalignment='bottom', horizontalalignment='center', fontsize=10))
 
     def add_total(self, value, title: Optional[str] = "Total"):
-        self.data = self.data.append(QFSeries([value], [title]))
-        self.total_value = self.data.index[-1]
+        series = QFSeries([value], [title])
+        self.add_decorator(DataElementDecorator(series))
+        self.total_value = series.index
 
     def flag_total(self, value):
         self.total_value = value
 
     def apply_data_element_decorators(self, data_element_decorators: List["DataElementDecorator"]):
-        pass
+        index = 0
+        for data_element in data_element_decorators:
```

Comment:
instead of incrementing index let's use enumerate

Reply:
In this case we need to track the index over two for loops so the enumerate will not work